### PR TITLE
Add methods that take Merlin transcript as param to VRF

### PR DESCRIPTION
**Pull request by Harrm**
Took changes from @iceseer 's fork, refactored a bit and added test. 

The goal is to allow passing a custom external transcript as a param in VRF functions.

### CODE REVIEW

File: test/sr25519/transcript.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    append_message(label, tmp);
+  }
+
+  auto data() {
+    return strobe_.data();
+  }
+
+  auto state() const {
+    return strobe_.state();
+  }
```
Comment (iceseer) [inline test/sr25519/transcript.hpp:77]: This method can be removed. It was added for debugging purpose. 'state' in strobe also.

File: src/sr25519.rs [hunk 2]
```diff
@@ -403,6 +404,105 @@ pub unsafe extern "C" fn sr25519_vrf_verify(
...
+    let raw_out_bytes = io.make_bytes::<[u8; SR25519_VRF_RAW_OUTPUT_SIZE as usize]>(BABE_VRF_PREFIX);
+    let check = u128::from_le_bytes(raw_out_bytes) < limit_int;
+
+    ptr::copy(io.to_output().as_bytes().as_ptr(), out_and_proof_ptr, SR25519_VRF_OUTPUT_SIZE as usize);
+    ptr::copy(proof.to_bytes().as_ptr(), out_and_proof_ptr.add(SR25519_VRF_OUTPUT_SIZE as usize), SR25519_VRF_PROOF_SIZE as usize);
+    if check {
+        VrfResult::create_val(true)
+    } else {
+        VrfResult::create_val(false)
+    }
```
Comment (iceseer) [inline src/sr25519.rs]: ```VrfResult::create_val(check)``` ?

File: src/sr25519.rs [hunk 3]
```diff
@@ -403,6 +404,105 @@ pub unsafe extern "C" fn sr25519_vrf_verify(
...
+        match public_key.vrf_verify(transcript, &given_out, &given_proof) {
+            Ok(val) => val,
+            Err(err) => return VrfResult::create_err(&err)
+        };
+    let raw_output = in_out.make_bytes::<[u8; SR25519_VRF_RAW_OUTPUT_SIZE as usize]>(BABE_VRF_PREFIX);
+
+    let threshold = slice::from_raw_parts(threshold_ptr, SR25519_VRF_THRESHOLD_SIZE as usize);
+    let mut threshold_arr: [u8; SR25519_VRF_THRESHOLD_SIZE as usize] = Default::default();
+    threshold_arr.copy_from_slice(&threshold[0..SR25519_VRF_THRESHOLD_SIZE as usize]);
+    let threshold_int = u128::from_le_bytes(threshold_arr);
```
Comment (iceseer) [inline src/sr25519.rs]: Will it work?
```c++
let threshold_int = u128::from_le_bytes(&threshold[0..SR25519_VRF_THRESHOLD_SIZE as usize]);

**Review by iceseer — CHANGES_REQUESTED** [review]

File: test/sr25519/transcript.hpp (hunk 1 above)
Comment (Harrm) [reply test/sr25519/transcript.hpp:77]: But I use it to initialize Strobe128 struct for bindings.

File: src/sr25519.rs (hunk 3 above)
Comment (Harrm) [reply src/sr25519.rs]: Nope, that would be an attempt to pass a reference to a slice, whereas u128::from_le_bytes requires a fixed-size array.

Comment (Harrm) [reply src/sr25519.rs]: You gave me a good idea though.

**Review by iceseer — APPROVED** [review]

**Review by kamilsa — APPROVED** [review]

**Pull request merged**